### PR TITLE
Revive remove default ghosting

### PR DESCRIPTION
**Pull request by lindsayad**
This eliminates libMesh's default ghosting for MOOSE simulations, and only adds ghosting when it is needed, e.g. when we have objects like:

- adaptivity
- DG
- interface kernels
- internal side user objects
- several output objects 
- `MultiAppProjectionTransfer`

and more in modules.

### CODE REVIEW

File: framework/src/actions/AdaptivityAction.C [hunk 1]
```diff
@@ -83,62 +86,122 @@ AdaptivityAction::AdaptivityAction(InputParameters params) : Action(params) {}
...
+                   getParam<bool>("recompute_markers_during_cycles"));
+
+    adapt.setPrintMeshChanged(getParam<bool>("print_changed_info"));
+
+    const std::vector<std::string> & weight_names =
+        getParam<std::vector<std::string>>("weight_names");
+    const std::vector<Real> & weight_values = getParam<std::vector<Real>>("weight_values");
+
+    int num_weight_names = weight_names.size();
+    int num_weight_values = weight_values.size();
```
Comment (permcody) [inline framework/src/actions/AdaptivityAction.C]: What's with the `int`s? Shouldn't this be `std::size_t` or `auto`?

File: framework/src/actions/AdaptivityAction.C [hunk 2]
```diff
@@ -83,62 +86,122 @@ AdaptivityAction::AdaptivityAction(InputParameters params) : Action(params) {}
...
-      weights[system.getVariable(0, name).number()] = value;
-    }
+      // If weights have been specified then set the default weight to zero
+      std::vector<Real> weights(system.nVariables(), 0);
 
-    std::vector<FEMNormType> norms(system.nVariables(), H1_SEMINORM);
+      for (int i = 0; i < num_weight_names; i++)
+      {
+        std::string name = weight_names[i];
+        double value = weight_values[i];
```
Comment (permcody) [inline framework/src/actions/AdaptivityAction.C]: `double`?

File: framework/src/actions/AdaptivityAction.C [hunk 3]
```diff
@@ -83,62 +86,122 @@ AdaptivityAction::AdaptivityAction(InputParameters params) : Action(params) {}
...
+        mooseError("Number of weight_names must be equal to number of weight_values in "
+                   "Execution/Adaptivity");
 
-      weights[system.getVariable(0, name).number()] = value;
-    }
+      // If weights have been specified then set the default weight to zero
+      std::vector<Real> weights(system.nVariables(), 0);
 
-    std::vector<FEMNormType> norms(system.nVariables(), H1_SEMINORM);
+      for (int i = 0; i < num_weight_names; i++)
```
Comment (permcody) [inline framework/src/actions/AdaptivityAction.C]: `int`?

File: framework/src/actions/AddPeriodicBCAction.C [hunk 4]
```diff
@@ -146,66 +148,103 @@ AddPeriodicBCAction::autoTranslationBoundaries()
...
-
-    // Add the pair of periodic boundaries to the dof map
-    nl.dofMap().add_periodic_boundary(pb, ipb);
-    if (displaced_problem)
-      displaced_problem->nlSys().dofMap().add_periodic_boundary(pb, ipb);
+    if (rm_params.areAllRequiredParamsValid())
+    {
+      auto rm_obj =
+          _factory.create<RelationshipManager>("ElementSideNeighborLayers",
+                                               "periodc_bc_ghosting" + Moose::stringify(count++),
```
Comment (permcody) [inline framework/src/actions/AddPeriodicBCAction.C]: I'm not sure I love this `count` business. Why not use the `for_whom` field here instead? I don't think we need to add more than one of these per field.

File: framework/src/actions/AddRelationshipManager.C [hunk 5]
```diff
@@ -26,13 +26,29 @@ AddRelationshipManager::AddRelationshipManager(InputParameters params) : Action(
...
+    _app.attachRelationshipManagers(Moose::RelationshipManagerType::GEOMETRIC);
+  }
+
+  if (_current_task == "attach_algebraic_rm")
+  {
+    const auto & all_action_ptrs = _awh.allActionBlocks();
+    for (const auto & action_ptr : all_action_ptrs)
+      action_ptr->addRelationshipManagers(Moose::RelationshipManagerType::ALGEBRAIC);
+
+    _app.attachRelationshipManagers(Moose::RelationshipManagerType::ALGEBRAIC);
```
Comment (permcody) [inline framework/src/actions/AddRelationshipManager.C]: Funny - this manual unroll created some code duplication. I think I liked it the other way better.

File: framework/src/auxkernels/GhostingAux.C [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+{
+  if (isNodal())
+    mooseError("GhostingAux only works on elemental fields.");
+
+  // Can only work on Replicated meshes
+  _mesh.errorIfDistributedMesh("GhostingAux");
+}
+
+void
+GhostingAux::initialize()
```
Comment (permcody) [inline framework/src/auxkernels/GhostingAux.C]: Still don't love this, but I guess that ship has sailed.

File: framework/src/auxkernels/GhostingAux.C [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+GhostingAux::GhostingAux(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _pid(getParam<processor_id_type>("pid")),
+    _functor_type(getParam<MooseEnum>("functor_type"))
+{
+  if (isNodal())
+    mooseError("GhostingAux only works on elemental fields.");
+
+  // Can only work on Replicated meshes
+  _mesh.errorIfDistributedMesh("GhostingAux");
```
Comment (permcody) [inline framework/src/auxkernels/GhostingAux.C]: This _really_ should be made to work with DistributedMesh. Figuring out the ghosting on ReplicatedMesh feels more like a testing exercise. Is this a MOOSE utility or not? This object is adding a new API to AuxKernel so it really should be first class IMHO.

File: framework/src/base/Moose.C [hunk 8]
```diff
@@ -201,6 +203,8 @@ addActionTypes(Syntax & syntax)
   registerTask("setup_function_complete", false);
   registerTask("setup_variable_complete", false);
   registerTask("ready_to_init", true);
+  appendMooseObjectTask("ready_to_init", InterfaceKernel);
```
Comment (permcody) [inline framework/src/base/Moose.C]: I'd prefer if these lines were moved up and clustered with existing `ready_to_init` tasks in the section up above.

File: framework/src/base/MooseApp.C [hunk 9]
```diff
@@ -1687,10 +1705,65 @@ MooseApp::addRelationshipManager(std::shared_ptr<RelationshipManager> relationsh
 }
 
 void
-MooseApp::attachRelationshipManagers(Moose::RelationshipManagerType rm_type)
+MooseApp::attachRelationshipManagers(const Moose::RelationshipManagerType & rm_type)
```
Comment (permcody) [inline framework/src/base/MooseApp.C]: This is an enum. Not sure we need to pass it by reference.

File: framework/src/relationshipmanagers/ProxyRelationshipManager.C [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  if (useDisplacedMesh())
+    info = info + " on displaced";
+
+  return info;
+}
+
+bool
+ProxyRelationshipManager::operator==(const RelationshipManager & /*rhs*/) const
+{
+  // There isn't a need to determine these because only the correct ones will be added
```
Comment (permcody) [inline framework/src/relationshipmanagers/ProxyRelationshipManager.C]: There's confidence!

File: framework/src/utils/Conversion.C [hunk 11]
```diff
@@ -346,25 +346,32 @@ stringToEnum<RelationshipManagerType>(const std::string & s)
...
-    case RelationshipManagerType::GEOMETRIC:
-      return "GEOMETRIC";
-    case RelationshipManagerType::ALGEBRAIC:
-      return "ALGEBRAIC";
-    case RelationshipManagerType::COUPLING:
-      return "COUPLING";
-    default:
-      return "ERROR";
-  }
+  // Connot make a switch statement because the boolean logic doesn't work well with the class type
```
Comment (permcody) [inline framework/src/utils/Conversion.C]: Cannot

File: modules/phase_field/src/auxkernels/FeatureFloodCountAux.C [hunk 12]
```diff
@@ -39,6 +39,10 @@ validParams<FeatureFloodCountAux>()
                              "(UNIQUE_REGION and VARIABLE_COLORING are nodal, CENTROID is "
                              "elemental, default: UNIQUE_REGION)");
 
+  params.addRelationshipManager("ElementSideNeighborLayers",
```
Comment (permcody) [inline modules/phase_field/src/auxkernels/FeatureFloodCountAux.C]: This object doesn't need a relationship manager. It operates on local elements only. Now the FeatureFloodCount UO *does* need an RM.

File: modules/phase_field/src/postprocessors/GrainTracker.C [hunk 13]
```diff
@@ -52,9 +52,18 @@ validParams<GrainTracker>()
...
   params += validParams<GrainTrackerInterface>();
 
-  params.registerRelationshipManagers("GrainTrackerHaloRM ElementPointNeighbors",
-                                      "GEOMETRIC ALGEBRAIC");
-  params.addPrivateParam<unsigned short>("element_point_neighbor_layers", 1);
+  params.addRelationshipManager(
+      "ElementPointNeighborLayers",
+      Moose::RelationshipManagerType::GEOMETRIC,
+
+      [](const InputParameters & obj_params, InputParameters & rm_params) {
```
Comment (permcody) [inline modules/phase_field/src/postprocessors/GrainTracker.C]: 👍

**Review by permcody — COMMENTED** [review]
I don't see any showstoppers here. You should probably address comments and then let's plan to merge this in and fixup apps.

File: framework/src/base/Moose.C (hunk 8 above)
Comment (lindsayad) [reply framework/src/base/Moose.C]: Huh? The only three occurrences of `ready_to_init` are right here.

Comment (permcody) [reply framework/src/base/Moose.C]: oh weird... I'm not sure if this is a bug or not. Note the name of the method used here. It's called "appendMooseObjectTask", which is supposed to correspond with "registerMooseObjectTask". The first registration or `ready_to_init` is not a MOOSE object task so I hope that the code is doing the right thing. Note that this is definitely a unique combination of these registration functions so it should probably be checked.

**Review by permcody — APPROVED** [review]
Approved with trepidation

**Pull request merged**